host/xtest/pkcs11_1000.c

```diff
@@ -3741,3 +3741,428 @@ static void xtest_pkcs11_test_1016(ADBG_Case_t *c)
+	};
+	CK_ATTRIBUTE parent_template_wo_derive[] = {
+		{ CKA_SENSITIVE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_EXTRACTABLE, &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) },
+		{ CKA_VALUE_LEN, &(CK_ULONG){16}, sizeof(CK_ULONG) },
+	};
+	CK_ATTRIBUTE parent_template_w_enc_der[] = {
+		{ CKA_VALUE_LEN, &(CK_ULONG){16}, sizeof(CK_ULONG) },
+		{ CKA_ENCRYPT, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_DERIVE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+	};
+
+	rv = init_lib_and_find_token_slot(&slot);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		return;
+
+	rv = init_test_token(slot);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_lib;
+
+	rv = init_user_test_token(slot);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_lib;
+
+	rv = C_OpenSession(slot, session_flags, NULL, 0, &session);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_lib;
+
+	/*
+	 * Parent AES Key
+	 * SENSITIVE = TRUE, EXTRACTABLE = FALSE
+	 * ALWAYS_SENSITIVE = TRUE, NEVER_EXTRACTABLE = TRUE
+	 */
+	rv = C_GenerateKey(session, &cktest_aes_keygen_mechanism,
+			   parent_template1, ARRAY_SIZE(parent_template1),
+			   &aes_key1);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_session;
+
+	/*
+	 * Parent AES Key
+	 * SENSITIVE = FALSE, EXTRACTABLE = TRUE
+	 * ALWAYS_SENSITIVE = FALSE, NEVER_EXTRACTABLE = FALSE
+	 */
+	rv = C_GenerateKey(session, &cktest_aes_keygen_mechanism,
+			   parent_template2, ARRAY_SIZE(parent_template2),
+			   &aes_key2);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_session;
+
+	Do_ADBG_BeginSubCase(c, "Derive Generic secret - AES-ECB Mechanism");
+
+	/*
+	 * Use AES key 1 as Parent key
+	 * 1. VALUE_LEN attribute not given in derivation template. Length
+	 * of key should be same as that of data length.
+	 * 2. Derivation template has SENSITIVE = TRUE, EXTRACTABLE = FALSE
+	 * Parent key has ALWAYS_SENSITIVE = TRUE, NEVER_EXTRACTABLE = TRUE
+	 * So derived key, ALWAYS_SENSITIVE will be same as SENSITIVE and
+	 * NEVER_EXTRACTABLE will be opposite of EXTRACTABLE
+	 * 3. LOCAL should be false
+	 */
+	datalen = 512;
+	keylen = 0;
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB_ENCRYPT_DATA,
+			    datalen, &derv_key_hdl, keylen, CKO_SECRET_KEY,
+			    CKK_GENERIC_SECRET, CK_TRUE, CK_FALSE);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_GetAttributeValue(session, derv_key_hdl, get_template,
+				 ARRAY_SIZE(get_template) - 1);
+	if (!ADBG_EXPECT_CK_OK(c, rv) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, secret_len, ==, datalen) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_class, ==, CKO_SECRET_KEY) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_keytype, ==,
+					  CKK_GENERIC_SECRET) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_local, ==, CK_FALSE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_sensitive, ==, CK_TRUE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_extract, ==, CK_FALSE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_asensitive, ==, CK_TRUE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_nextract, ==, CK_TRUE))
+		goto out;
+
```

Comment:
Could we somehow verify that the derived key is correct in here?

Reply:
I checked with offline tools for an AES ECB case by dumping the original generated key, derived key and performing encryption of the parent key using the algo. Didn't find any elegant way to add it here so just checked success of operation.

Reply:
Agree. We should test per algo derivation somewhere else. Here we're focusing on basics and attributes value.